src/tls.cpp

```diff
@@ -255,35 +264,60 @@ void tls_exit(void) {
     }
 }
 
-void *tls_connect(int fd) {
+int tls_connect_or_accept(SSL* ssl) {
+    int ret = 0;
+    bool try_again = false;
+    auto ssl_func = (SSL_is_server(ssl) ? SSL_accept : SSL_connect);
+    
+    do {
+        try_again = false;
+        ret = ssl_func(ssl);
+        if (ret < 0 && !g_b_exit) {
```

Comment:
May be `ret <=0`?

Reply:
We do the loop only in case of WANT_READ/WRITE for which -1 is returned. 0 can be returned in case of socket/tls shutdown , for which we do not want to enter the accept/connect loop. Because of that it has to be <0.

Reply:
So, getting 0 and checking SSL_get_error() you can break on else with tls_log_err() too.
But may be it is ok to do as you.

Reply:
Going to check ret <= 0